### ISSUE TITLE
Disconnect the Session from the Connection

### ISSUE THREAD

**Issue by hawkowl** [wamp-proto/wamp-proto#260]
Advanced Profile WAMP brokers may want to implement some form of 'lasting session' for clients. Upon resumption of the session, it will still have all the existing Subscriptions and registrations, although it will not receive messages sent to those subscriptions nor method calls to those registrations that were sent offline -- that is up to the client to fetch using the publication history meta API.

**Comment by oberstet** [comment]
Pls see the sketch here https://github.com/wamp-proto/wamp-proto/blob/master/rfc/session-state-machine.md

**Comment by hawkowl** [comment]
@oberstet yeah that looks exactly like what I was thinking too... should this belong in the advanced profile somewhere?

**Comment by hawkowl** [comment]
@oberstet what's missing from the sketch? I think we need:

- Clarification about reconnecting to a session that Crossbar thinks isn't paused (for example, long TCP timeouts on mobile stuff)
- How a client makes a session paused (a `GOODBYE` with a pause item in the details dict?)
- Clarification about what happens when a router can't let a client join a session (e.g. it is a different router than what the WAMP session was opened on), what the `WELCOME` looks like (I am assuming either no `Welcome.Details.resumed` if the server doesn't support it, or `Welcome.Details.resumed := false`
- Destroying resumable sessions -- explicitly through `GOODBYE`, allowances for the router to time-out sessions?

I think there's enough here that I could try implementing the session resumption in Autobahn|Python and see how it goes?

**Comment by oberstet** [comment]
> what's missing from the sketch?

Rereading my sketch, it seems like I was interrupted in the middle of the last sentence;) I can't remember why. Well.

Yes, at least the specific things you raise would need treatment. Likely there are more nitty gritty details we'll stumble over when implementing.

> I think there's enough here that I could try implementing the session resumption in Autobahn|Python and see how it goes?

Cool!!

In the mid-term, we need to get consensus/feedback from other WAMP implementors, but I am fine with first incubating this and more so: gain a little bit implementation experience. This spec/implementation co-design approach worked fine for WAMP historically. It's kind of a balancing act.

What I also think is: we should make sure that whatever we do for session resumption within AutobahnPython fits together with the "new API" stuff.

This is the chance to come up with something that really moves AB to the "final level" (1.0) rgd WAMP.

Session resumption and "new API" are _the_ two most important pieces missing for 1.0 (besides full PEP8 .. public API naming).

So we definitely should take @meejah on board with this .. he has done a lot of work on "new API".

**Comment by hawkowl** [comment]
@oberstet this is next on my list to approach... did you agree with the things I noted, should I think up solutions and add them to the sketch, or did you want to do that? I think the first one (reconnecting to a session that Crossbar thinks isn't paused) is going to be a difficult one (since that involves kicking off an existing session if the resuming transport is prioritised over the existing one). Or whether we leave it up to implementations how they handle resuming existing sessions, although I think the RFC should really have 'one way' for doing it...

I'll see what I can implement -- the message-level stuff needs to happen before anything else anyway, then I think the CB impl, and then the new API? I'm also wondering how this will work in AB -- I am kind of figuring it would be automatic in the background for connection loss/resumption, but do we want to expose it to users directly?

**Comment by oberstet** [comment]
@hawkowl I think best would be to get rid of the Markdown file and transfer the content to a PR. Put in additions or changes as you see fit. Then we can work from there.

> Clarification about reconnecting to a session that Crossbar thinks isn't paused (for example, long TCP timeouts on mobile stuff)
 
I'd suggest: "A client that connects, wishing to resume a session the Router already has collected and destroyed will get sent a `GOODBYE` message with a specific `reason`." or something like this. Such a client can then retry without resuming, starting a completely new session.

> How a client makes a session paused (a GOODBYE with a pause item in the details dict?)

"pause" item .. sounds good to me! 

> Clarification about what happens when a router can't let a client join a session (e.g. it is a different router than what the WAMP session was opened on), what the WELCOME looks like (I am assuming either no Welcome.Details.resumed if the server doesn't support it, or Welcome.Details.resumed := false

Similar to above .. send a `GOODBYE` with specific `reason`.

> Destroying resumable sessions -- explicitly through GOODBYE, allowances for the router to time-out sessions?

The time out of sessions should be under control of the router administrator, since it affects resource consumption. The router configuration is implementation specific, so I'd say, this isn't something for the spec, but of course we need it in Crossbar.io (via a usual config knob).

Explicit destruction of session: a flag set by a Client in the initial `GOODBYE` it send when leaving, yes. Perfect.

**Comment by hawkowl** [comment]
I did some stuff up in a PR in RFC form, I'll get rid of the .md in it.

**Comment by oberstet** [comment]
An alternative (nicer) behavior for the Router: a Client that set `resume=true`, and where the Router cannot resume the session, will answer by entering the usual `CHALLENGE/AUTHENTICATE` cycle, or will `WELCOME`, with `was_resumed=false` or something ..

**Comment by oberstet** [comment]
There is a draft in this PR https://github.com/wamp-proto/wamp-proto/pull/264, and also lots of more aspects in the discussion in this PR
